static/js/components/Editor.js

```diff
@@ -0,0 +1,310 @@
+// @flow
+import React from "react"
+import R from "ramda"
+import { connect } from "react-redux"
+import { EditorState } from "prosemirror-state"
+import { EditorView } from "prosemirror-view"
+import { exampleSetup } from "prosemirror-example-setup"
+import {
+  schema,
+  defaultMarkdownParser,
+  defaultMarkdownSerializer
+} from "prosemirror-markdown"
+import { toggleMark, wrapIn } from "prosemirror-commands"
+import { wrapInList } from "prosemirror-schema-list"
+
+import AddLinkMenu from "../components/AddLinkMenu"
+
+import { makeUUID } from "../lib/util"
+import { configureForm } from "../lib/forms"
+
+import type { Dispatch } from "redux"
+
+// helpers for the menu bar
+type MenuBarItem = {
+  command: Function,
+  title: string,
+  icon: React$Element<"i">,
+  mark?: Object
+}
+
+// each array in this array represents a 'group' of buttons
+// (we want to group certain buttons together with a border
+// around them in the UI)
+export const menuBarManifest: Array<Array<MenuBarItem>> = [
+  [
+    {
+      command: toggleMark(schema.marks.strong),
+      title:   "Bold",
+      icon:    <i className="material-icons format_bold">format_bold</i>,
+      mark:    schema.marks.strong
+    }
+  ],
+  [
+    {
+      command: toggleMark(schema.marks.em),
+      title:   "Italic",
+      icon:    <i className="material-icons format_italic">format_italic</i>,
+      mark:    schema.marks.em
+    }
+  ],
+  [
+    {
+      command: wrapInList(schema.nodes.bullet_list),
+      title:   "bullet list",
+      icon:    (
+        <i className="material-icons format_list_bulleted">
+          format_list_bulleted
+        </i>
+      )
+    },
+    {
+      command: wrapInList(schema.nodes.ordered_list),
+      title:   "ordered list",
+      icon:    (
+        <i className="material-icons format_list_numbered">
+          format_list_numbered
+        </i>
+      )
+    },
+    {
+      command: wrapIn(schema.nodes.blockquote),
+      title:   "quote",
+      icon:    <i className="material-icons format_quote">format_quote</i>
+    }
+  ]
+]
+
+export const newLinkForm = () => ({
+  url: ""
+})
+
+type Props = {
+  onChange: (serialized: string) => void,
+  initialValue?: string,
+  autoFocus?: boolean,
+  dispatch: Dispatch<*>
+}
+
+type State = {
+  editorState: Object,
+  showLinkMenu: boolean
+}
+
+export class Editor extends React.Component<Props, State> {
+  node: { current: null | React$ElementRef<typeof HTMLDivElement> }
+  view: Object
+  uuid: string
+  formHelpers: Object
+  dispatchTransaction: Function
+
+  constructor(props: Props) {
+    super(props)
+    this.node = React.createRef()
+    this.uuid = makeUUID(10)
+
+    const { getForm, actionCreators } = configureForm(this.uuid, newLinkForm)
+    this.formHelpers = {
+      getForm,
+      ...actionCreators
+    }
```

Comment:
yeah, the `uuid` is so that we can have more than one editor component on the page at the same time without them trying to edit the same form. unfortunately that means I need to calculate the `formHelpers` and so on dynamically at runtime, so I can't use the `mapDispatchToProps` pattern here. in general I think it's best not to even have a constructor, but in cases like this component there are things that have to be done there (like setting up refs and stuff like that).